File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/factions.ts
/**
 * Campaign-scoped factions and bounded per-character reputation history.
 * Reputation totals are clamped to [-100,100]; each accepted change appends
 * an immutable history record so the DM (and the affected player) can audit
 * how a total was reached.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import {
  authenticate,
  isActor,
  isApiResult,
  findCampaign,
  findMemberByCharacterId,
  requireParticipant,
} from './shared.ts';

type FactionRow = {
  campaign_id: string;
  faction_id: string;
  name: string;
};

type ReputationHistoryRow = {
  faction_id: string;
  character_id: string;
  reputation: number;
  delta: number;
  reason: string;
};

const FACTION_NOT_FOUND: ApiResult = { status: 404, body: { error: 'faction not found' } };
const REPUTATION_MIN = -100;
const REPUTATION_MAX = 100;

function findFaction(db: ReturnType<typeof getDb>, campaignId: string, factionId: string): FactionRow | ApiResult {
  const faction = db
    .prepare('SELECT campaign_id, faction_id, name FROM play_campaign_factions WHERE campaign_id = ? AND faction_id = ?')
    .get(campaignId, factionId) as FactionRow | undefined;
  return faction ?? FACTION_NOT_FOUND;
}

function nextReputationSequence(db: ReturnType<typeof getDb>, campaignId: string, factionId: string): number {
  const row = db
    .prepare(
      'SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_faction_reputation_history WHERE campaign_id = ? AND faction_id = ?',
    )
    .get(campaignId, factionId) as { max_sequence: number };
  return row.max_sequence + 1;
}

export function createCampaignFaction(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may create factions' } };
  }

  const factionId = body.faction_id;
  if (typeof factionId !== 'string' || factionId.length === 0) {
    return { status: 400, body: { error: 'faction_id must be a non-empty string' } };
  }

  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const existing = findFaction(db, campaignId, factionId);
  if (!isApiResult(existing)) {
    return { status: 409, body: { error: 'faction_id already exists in this campaign' } };
  }

  db.prepare('INSERT INTO play_campaign_factions (campaign_id, faction_id, name) VALUES (?, ?, ?)').run(
    campaignId,
    factionId,
    name,
  );

  return { status: 201, body: { faction_id: factionId, name } };
}

export function adjustFactionReputation(
  authHeader: string | undefined,
  campaignId: string,
  factionId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may change reputation' } };
  }

  const faction = findFaction(db, campaignId, factionId);
  if (isApiResult(faction)) return faction;

  const characterId = body.character_id;
  if (typeof characterId !== 'string' || characterId.length === 0) {
    return { status: 400, body: { error: 'character_id must be a non-empty string' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) {
    return { status: 400, body: { error: 'character_id must identify a campaign member character' } };
  }

  const delta = body.delta;
  if (typeof delta !== 'number' || !Number.isInteger(delta) || delta === 0 || delta < -25 || delta > 25) {
    return { status: 400, body: { error: 'delta must be a nonzero integer in [-25,25]' } };
  }

  const reason = body.reason;
  if (typeof reason !== 'string' || reason.length === 0) {
    return { status: 400, body: { error: 'reason must be a non-empty string' } };
  }

  const current = db
    .prepare(
      'SELECT reputation FROM play_campaign_faction_reputation WHERE campaign_id = ? AND faction_id = ? AND character_id = ?',
    )
    .get(campaignId, factionId, characterId) as { reputation: number } | undefined;

  const currentTotal = current?.reputation ?? 0;
  const nextTotal = Math.max(REPUTATION_MIN, Math.min(REPUTATION_MAX, currentTotal + delta));

  db.prepare(
    `INSERT INTO play_campaign_faction_reputation (campaign_id, faction_id, character_id, reputation)
     VALUES (?, ?, ?, ?)
     ON CONFLICT (campaign_id, faction_id, character_id) DO UPDATE SET reputation = excluded.reputation`,
  ).run(campaignId, factionId, characterId, nextTotal);

  const sequence = nextReputationSequence(db, campaignId, factionId);
  db.prepare(
    `INSERT INTO play_campaign_faction_reputation_history
       (campaign_id, faction_id, sequence, character_id, reputation, delta, reason)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
  ).run(campaignId, factionId, sequence, characterId, nextTotal, delta, reason);

  return {
    status: 201,
    body: { faction_id: factionId, character_id: characterId, reputation: nextTotal, delta, reason },
  };
}

export function getFactionReputation(authHeader: string | undefined, campaignId: string, factionId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const faction = findFaction(db, campaignId, factionId);
  if (isApiResult(faction)) return faction;

  const rows = db
    .prepare(
      `SELECT faction_id, character_id, reputation, delta, reason
       FROM play_campaign_faction_reputation_history
       WHERE campaign_id = ? AND faction_id = ?
       ORDER BY sequence ASC`,
    )
    .all(campaignId, factionId) as ReputationHistoryRow[];

  let entries = rows;
  if (actor.username !== campaign.owner) {
    const ownCharacterId = db
      .prepare('SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?')
      .get(campaignId, actor.username) as { character_id: string } | undefined;
    entries = ownCharacterId ? rows.filter((row) => row.character_id === ownCharacterId.character_id) : [];
  }

  return {
    status: 200,
    body: {
      faction_id: factionId,
      entries: entries.map((entry) => ({
        faction_id: entry.faction_id,
        character_id: entry.character_id,
        reputation: entry.reputation,
        delta: entry.delta,
        reason: entry.reason,
      })),
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/factions.rb
# Campaign-scoped factions and bounded character reputation history for
# the /v1/play surface. The DM creates factions and records reputation
# changes; each change is clamped into [-100, 100] and appended to an
# immutable per-faction/character history.

def find_play_faction!(campaign_id, faction_id)
  faction = db.execute(
    'SELECT * FROM play_factions WHERE campaign_id = ? AND faction_id = ?',
    [campaign_id, faction_id]
  ).first
  halt 404, { error: 'faction not found' }.to_json unless faction
  faction
end

def next_play_faction_reputation_sequence(campaign_id, faction_id)
  db.execute(
    'SELECT COALESCE(MAX(sequence), 0) + 1 AS n FROM play_faction_reputation_history WHERE campaign_id = ? AND faction_id = ?',
    [campaign_id, faction_id]
  ).first['n']
end

def play_faction_reputation_entry(row)
  {
    faction_id: row['faction_id'],
    character_id: row['character_id'],
    reputation: row['reputation'],
    delta: row['delta'],
    reason: row['reason']
  }
end

post '/v1/play/campaigns/:id/factions' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  faction_id = body['faction_id']
  name = body['name']

  halt 400, { error: 'invalid faction_id' }.to_json unless faction_id.is_a?(String) && !faction_id.empty?
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?

  existing = db.execute(
    'SELECT 1 FROM play_factions WHERE campaign_id = ? AND faction_id = ?',
    [campaign['id'], faction_id]
  ).first
  halt 409, { error: 'faction_id already exists' }.to_json if existing

  db.execute(
    'INSERT INTO play_factions (campaign_id, faction_id, name) VALUES (?, ?, ?)',
    [campaign['id'], faction_id, name]
  )

  status 201
  { faction_id: faction_id, name: name }.to_json
end

post '/v1/play/campaigns/:id/factions/:faction_id/reputation' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  faction = find_play_faction!(campaign['id'], params[:faction_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  character_id = body['character_id']
  delta = body['delta']
  reason = body['reason']

  halt 400, { error: 'invalid character_id' }.to_json unless character_id.is_a?(String) && !character_id.empty?

  member = db.execute(
    'SELECT 1 FROM play_campaign_members WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], character_id]
  ).first
  halt 400, { error: 'unknown character' }.to_json unless member

  halt 400, { error: 'invalid delta' }.to_json unless integerish(delta) && delta.to_i != 0 && delta.to_i.between?(-25, 25)
  delta = delta.to_i

  halt 400, { error: 'invalid reason' }.to_json unless reason.is_a?(String) && !reason.empty?

  current_row = db.execute(
    'SELECT reputation FROM play_faction_reputation WHERE campaign_id = ? AND faction_id = ? AND character_id = ?',
    [campaign['id'], faction['faction_id'], character_id]
  ).first
  current = current_row ? current_row['reputation'] : 0

  reputation = [[current + delta, -100].max, 100].min

  db.execute(
    'INSERT INTO play_faction_reputation (campaign_id, faction_id, character_id, reputation) VALUES (?, ?, ?, ?) ' \
    'ON CONFLICT (campaign_id, faction_id, character_id) DO UPDATE SET reputation = excluded.reputation',
    [campaign['id'], faction['faction_id'], character_id, reputation]
  )

  sequence = next_play_faction_reputation_sequence(campaign['id'], faction['faction_id'])
  db.execute(
    'INSERT INTO play_faction_reputation_history ' \
    '(campaign_id, faction_id, character_id, sequence, reputation, delta, reason) VALUES (?, ?, ?, ?, ?, ?, ?)',
    [campaign['id'], faction['faction_id'], character_id, sequence, reputation, delta, reason]
  )

  status 201
  { faction_id: faction['faction_id'], character_id: character_id, reputation: reputation, delta: delta, reason: reason }.to_json
end

get '/v1/play/campaigns/:id/factions/:faction_id/reputation' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  faction = find_play_faction!(campaign['id'], params[:faction_id])

  rows = db.execute(
    'SELECT * FROM play_faction_reputation_history WHERE campaign_id = ? AND faction_id = ? ORDER BY sequence ASC',
    [campaign['id'], faction['faction_id']]
  )

  unless is_owner
    own_member = db.execute(
      'SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?',
      [campaign['id'], user['username']]
    ).first
    own_character_id = own_member ? own_member['character_id'] : nil
    rows = rows.select { |row| row['character_id'] == own_character_id }
  end

  { faction_id: faction['faction_id'], entries: rows.map { |row| play_faction_reputation_entry(row) } }.to_json
end
